### PR TITLE
Another attempt to make layout more reliable

### LINKED ISSUE #4449: Ensure canvas state reset at start of drawing
Continuing discussion from #4405. The [`render()` method](https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/models/plots/plot_canvas.coffee#L374) in `plot_canvas.coffee` is the main entry point for the rendering; we should assume a clear state of the context because at the start of that function, we [make a call](https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/models/plots/plot_canvas.coffee#L407-L411) to `@canvas_view.prepare_canvas()`, which might resize the canvas (which resets context state).

All our `ctx.save()` and `ctx.restore()` usage pairs up as it should, however, if an error occurs (for whatever reason), any matching `ctx.restore()` may not be called, leaving the context in an unpredictable state. Since one of the first things we do in `plot_canvas.render()` is set a scaling transformation for the HIDPI, the plot will generally "blow up" in such a situation.

Although this is not a terrible thing per-see, as it only occurs when something is going wrong anyway, I think it can be a nasty / distracting side-effect. People might be wondering (and reporting) the plot from blowing up while the error is due to something completely unrelated.

Therefore, I propose to _ensure_ that the state of the context is clear before we start rendering, by making it reset in the `@canvas_view.prepare_canvas()` method (which already does a reset when it resizes the canvas).

Since there is no `ctx.reset()`, I propose to call `ctx.restore()` a couple of times. I agree that the _number_ of times to call `ctx.restore()` is rather arbitrary, and it generally means that we'll be doing a few unnecessary calls to `ctx.restore()` but I don't think the effect would be significant. 

A more sophisticated approach might be to keep tack of the depth of the stack by replacing `ctx.save()` and ctx.restore()` and updating a counter, and resetting based on that, but that feels a bit hacky.

### LINKED ISSUE #5967: Unexpected behavior when deleting from layout.children
I have encountered a bug when updating row contained in a column.

When `r1.children[0]` is deleted, r2 disappears and only the tools are left.  I would expect a column with `s1` above `s2`.  Is deleting from the children array not supported?

Example IPython notebook [https://github.com/paddymul/bokeh_bugs/blob/master/Bokeh%20row%20update%20artifact.ipynb](url)

```
#this messes up the rendering row2
r1 = row(s1,s2)
r2 = row(s3)
displayed_obj = column(r1, r2)

handle = show(displayed_obj, notebook_handle=True)
time.sleep(5)
del(r1.children[0])
push_notebook(handle=handle)
```
screen shot 2017-03-08 at 4 40 31 pm

Versions
python 2.7.11
IPython notebook 4.1.2
Bokeh 0.12.4

Thanks

### PR DESCRIPTION
**Pull request by mattpap**
- [x] implement per-root layout solver
- [ ] make layout reinitializable

fixes #5967 
? #5518
? #4810
fixes #4449

? #1650 -- not fixed because constraints are still bound to models, and there are other issues to resolve (e.g. we shouldn't touch `plot.renderers`).

### CODE REVIEW

**Comment by mattpap** [comment]
I just noticed that sometimes (e.g. in grid plots), layout enter an infinite loop of changing constraints, recomputing solution and triggering events. I'm not sure right now how much effort is needed to fix this.

**Comment by philippjfr** [comment]
I've just been testing this PR and comparing it to master. As far as I can tell this PR severely degrades performance in the layout solver. Notebooks containing a lot of plots now take considerably longer to render, pegging CPU load at 100% throughout and even once all the plots are done rendering, to the point that the notebook becomes unresponsive. 

When plotting column or gridplots I can now watch each subplot appear one by one when before they would all (at least appear to) render in one go. Specifically I tried the two main holoviews bokeh tutorials ([1](http://holoviews.org/Tutorials/Bokeh_Elements.html) and [2](http://holoviews.org/Tutorials/Bokeh_Backend.html)), which would take a little while to render before but basically freeze my browser tab when I try to regenerate them with this PR.

Sorry for the negative feedback and unfortunately I don't have any suggestions on how to fix/improve this either.

**Comment by bryevdv** [comment]
As a concrete example, the `models/file/iris_splom.py` example comes up with some of the subplots missing, taking a few seconds to show them.

**Comment by bryevdv** [comment]
I think there may be regressions in master too, though. If I run the "howto" LInked Paning notebook with master and  this PR, and repeatedly run the last three cells over and over, then the `show` gets progressively slower and worse. But it seems to get worse much faster on master

**Comment by mattpap** [comment]
I figured out the reason for performance regression. I'm working on a solution.

**Comment by bryevdv** [comment]
@mattpap can you confrim you are talking about a solver related and not the notebook show issue in #6037

**Comment by mattpap** [comment]
@bryevdv, no, I'm talking about regression introduced in this PR. Also, I'm working on a better fix, because what committed earlier is not a complete fix.

**Comment by philippjfr** [comment]
>Also, I'm working on a better fix, because what committed earlier is not a complete fix.

Just to say I can confirm it's no longer cripplingly slow now to the point of freezing my browser tab, but it's still considerably slower than before. I'll test again when there's a more complete fix.

**Comment by mattpap** [comment]
> `models/file/iris_splom.py`

This example is slow for another, unrelated reason. In this case, x and y ranges are shared across all plots and initialization of each next plot forces rerendering of all previously rendered plots (due to ranges being "updated").

**Comment by bryevdv** [comment]
I just merged the other PR for notebook slowdown. @philippjfr can you merge master into this PR locally, and see how the combination is?

**Comment by bryevdv** [comment]
@mattpap actually can you merge/rebase master?

**Comment by mattpap** [comment]
@bryevdv,  I think I resolved all performance regressions of this PR.

**Comment by bryevdv** [comment]
Making any selection in `crossfilter` blanks plot (no console messages)

screen shot 2017-03-22 at 15 23 01

**Comment by bryevdv** [comment]
Also get this error with lass select tool:
```
TypeError: this.throttled_render is not a function. (In 'this.throttled_render()', 'this.throttled_render' is undefined)
```

**Comment by mattpap** [comment]
> `TypeError: this.throttled_render is not a function. (In 'this.throttled_render()', 'this.throttled_render' is undefined)`

That's messed up. This shouldn't have ever worked. Fixed 477aafa.

**Comment by bryevdv** [comment]
Confirmed lasso select working now. Also all other apps seem to be functioning as expected. Checking notebooks next.

**Comment by mattpap** [comment]
> Making any selection in crossfilter blanks plot (no console messages)

Plot appears again when browser is resized or `resize()` method called. This is something trivial yet annoying.

**Comment by bryevdv** [comment]
notebook comms, notebook server embed and general notebook all appear to be working well on OSX.

**Comment by philippjfr** [comment]
Just tested the holoviews tutorials, performance even on large grids is now back to normal.

**Comment by mattpap** [comment]
btw. This is a very useful example I've been working with during work on this PR:
```
from bokeh.io import curdoc
from bokeh.plotting import figure
from bokeh.layouts import column
from bokeh.models.widgets import Select

def get_figure(color, n):
    p = figure(title="markers.py example")
    x = range(100)
    y = [ xi**n for xi in x ]
    p.line(x, y, color=color, name="color=%s, n=%s" % (color, n))
    return p

w = Select(options=['red', 'green', 'blue'], value='red')

f1 = get_figure('red', 1)
f2 = get_figure('red', 2)
f3 = get_figure('red', 3)
c = column(w, f1, f2, f3)

def update(name, old, new):
    c.children[1] = get_figure(w.value, 1)
    c.children[2] = get_figure(w.value, 2)
    c.children[3] = get_figure(w.value, 3)

"""
def update(name, old, new):
    c.children[1:] = [
        get_figure(w.value, 1),
        get_figure(w.value, 2),
        get_figure(w.value, 3),
    ]
"""

w.on_change('value', update)

curdoc().add_root(c)
```
The commented out code should work as well, but it's a much simpler and less useful case. This should be turned into a selenium tests.

**Comment by bryevdv** [comment]
OK, if the crossfilter example issue can be resolved, I'd support this plan:

* whatever needed (e.g. skipping) to get tests green
* more manual testing (by me) 
* merge and RC for tomorrow

**Comment by bryevdv** [comment]
> The commented out code should work as well, but it's a much simpler and less useful case. This should be turned into a selenium tests. 

Absolutely agree we need some selenium tests around apps at this point. I think we could run bokeh server on saucelabs? ping @canavandl

**Comment by mattpap** [comment]
> if the crossfilter example issue can be resolved

It must be. The above code is now broken as well (used to work earlier today).

**Comment by bryevdv** [comment]
> It must be. 
 
Obviously :) I was outlining the minimum requirements for a release this week.

**Comment by mattpap** [comment]
This can be changed to a non-server example as well. I've been working also with TypeScript examples as well, just then you won't see `PATCH-DOC` errors.

**Comment by mattpap** [comment]
I'm chasing my tail here. I fix one thing, another gets broken.

**Comment by bryevdv** [comment]
How can I help? Are there avenues I can help investigate?

File: bokehjs/src/coffee/models/layouts/layout_dom.coffee [hunk 1]
```diff
@@ -1,42 +1,128 @@
...
+  _reset_solver: () ->
+    if not @is_root
+      @parent._reset_solver()
+    else
+      @_solver.clear()
+      @_init_solver()
+
+  _init_solver: () ->
+    @_doc_width = new Variable("document_width")
+    @_doc_height = new Variable("document_height")
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/layouts/layout_dom.coffee]: for later: `_root_width` seems more descriptive

**Comment by bryevdv** [comment]
@mattpap my observation is that `prepare_canvas` is not getting called enough. In the interest of calling it more (but not all the time) I tried this:
```
diff --git a/bokehjs/src/coffee/models/plots/plot_canvas.coffee b/bokehjs/src/coffee/models/plots/plot_canvas.coffee
index 603db3240..01a5406d3 100644
--- a/bokehjs/src/coffee/models/plots/plot_canvas.coffee
+++ b/bokehjs/src/coffee/models/plots/plot_canvas.coffee
@@ -514,6 +514,9 @@ export class PlotCanvasView extends BokehView

   render: () ->
     logger.trace("PlotCanvas.render() for #{@model.id}")
+    if not @_canvas_prepared?
+      @canvas_view.prepare_canvas()
+      @_canvas_prepared = true

     if Date.now() - @interactive_timestamp < @model.plot.lod_interval
       if not @lod_started
```
This fixes the crossfilter example. I am checking others now, but so far so good. Would this be "good enough" for now?

**Comment by mattpap** [comment]
@bryevdv, this change is likely to break at last some forms of resizing. I have a different solution ready, but that in turn breaks toolbar positioning. There are so many things badly implemented that when I do one thing right, I need to do at least one other as well, and so on.

**Comment by mattpap** [comment]
Actually, I read this wrong.

**Comment by bryevdv** [comment]
@mattpap I just tried with all the various embed resizing and other resizing examples, and they seem to work as expected

**Comment by mattpap** [comment]
Lets proceed then.

**Comment by bryevdv** [comment]
@mattpap OK, can you skip or update whatever tests need to be? If not now, then in the morning? I will merge and get an RC out when it is green

**Comment by bryevdv** [comment]
@mattpap gmap not rendering until a resize, then reporting 
```
[Error] Error: detached element
```

**Comment by bryevdv** [comment]
Actually I think I had an old kernel running, after a restart all I can get for gmap is 
```
Javascript error adding output!
TypeError: undefined is not an object (evaluating 'window.google.maps')
See your browser Javascript console for more details.
```

**Comment by bryevdv** [comment]
so, FWIW the `CRITICAL` push notebook issue seemed to be fixed by the other hotfix PR for notebook perf regression. Given that, the problems remaining are problems that already existed. I think there is just too much risk to merge this for `0.12.5`. Let's get the other events PR in and this can be your primary focus for a (much smaller) `0.12.6` milestone.

**Comment by bryevdv** [comment]
FYI with fix above gmaps behave very badly on reloads, reshows. Basically `request_render` can get called before `initialize` somehow, then it falls over when `is_paused` is undefined.

**Comment by bryevdv** [comment]
I verified #5959 is working on master so I am closing it separately.

**Comment by mattpap** [comment]
> I think you are calling render earlier

Yes, I changed the (DOM) rendering model, so now there is no delayed or deferred rendering. This requires further changes in other places to make it reliable.

**Comment by mattpap** [comment]
> `[Error] Error: detached element`

This shouldn't happen, but can because we start rendering before root `el` is added to DOM. This is one of changes I referenced in the earlier comment.

**Comment by mattpap** [comment]
> and this can be your primary focus for a (much smaller) 0.12.6 milestone.

OK. This will allow to actually fix things.

**Comment by bryevdv** [comment]
Just FYI the examples under `howto/layout` do not look good/correct with this PR currently

**Comment by bryevdv** [comment]
I can confirm that the canvas is the correct size, the drawing is just 1/4 too small

**Comment by bryevdv** [comment]
@mattpap this it the fix for the HiDPI problem
```
diff --git a/bokehjs/src/coffee/models/plots/plot_canvas.coffee b/bokehjs/src/coffee/models/plots/plot_canvas.coffee
index e2329f5eb..f2da733f1 100644
--- a/bokehjs/src/coffee/models/plots/plot_canvas.coffee
+++ b/bokehjs/src/coffee/models/plots/plot_canvas.coffee
@@ -544,7 +544,7 @@ export class PlotCanvasView extends DOMView
     @model.frame._update_mappers()

     ctx = @canvas_view.ctx
-    ctx.pixel_ratio = ratio = @canvas_view.pixel_ratio  # Also store on cts for WebGL
+    ctx.pixel_ratio = ratio = @canvas.pixel_ratio  # Also store on cts for WebGL

     # Set hidpi-transform
     ctx.save()  # Save default state, do *after* getting ratio, cause setting canvas.width resets transforms
```

**Comment by bryevdv** [comment]
`examples/plotting/server/responsive.py` no longer trigger updates on resize

**Comment by mattpap** [comment]
> this is the fix for the HiDPI problem

Sure, this is what happens when the same value is stored all over the code.

**Comment by bryevdv** [comment]
FYI `plotting/server/responsive.py` still not firing events, probably just some missing plumbing

**Comment by bryevdv** [comment]
Unfortunately all the examples under `howto/layouts` are now seriously broken for me after the latest commits.

**Comment by mattpap** [comment]
> examples under howto/layouts are now seriously broken

None are tested on travis ci, so that's not unexpected. I'm investigating.

**Comment by canavandl** [comment]
The bokeh/tests/integration/interaction/test_datatable failure is real, when clicking on a cell to edit a ``Error: model of a view wasn't configured`` exception is raised.

**Comment by canavandl** [comment]
It appears that the ``sizing_mode='scale_width'`` isn't working, but the others are fine.

**Comment by mattpap** [comment]
> The bokeh/tests/integration/interaction/test_datatable

I already fixed that one locally.

**Comment by bryevdv** [comment]
Let's plan to discuss this in the morning. This PR is blocking several other avenues of work (due to the merges that would be extremely difficult to resolve). If this is still having problems by tomorrow AM CST then we will need to take a step back, let other work proceed, and figure out how to accomplish this in a series of smaller PRs. There is a fair amount of cleanup and incidental refactoring in this PR that could be done incrementally and verified, to reduce the size of the actual relevant changes that need review and evaluation.

**Comment by mattpap** [comment]
> plotting/server/responsive.py still not firing events, probably just some missing plumbing

Those events are intentionally disabled at this point to make debugging easier. Otherwise there is many "unable to apply patch" errors in JS console. Those errors are indicative of design issues in the server, but are otherwise harmless, at least to layout.

**Comment by mattpap** [comment]
> Unfortunately all the examples under howto/layouts are now seriously broken

I broke it when making adjustments to figure out what's wrong with unit tests. It's fixed now. I also enabled `howto/layouts` on travis ci.

**Comment by mattpap** [comment]
Given that tests should be all green at this point, I should be able to finish this PR tomorrow.

**Comment by mattpap** [comment]
One regression to note is that examples' tests run now 22 minutes instead of 16 (I'm excluding newly enabled examples). A lot more examples timeout in JS. After looking at the log, I suspect the recently added "finished rendering" heuristic may be at fault, because otherwise everything works as expected and there are no image diff failures (besides two spurious).

**Comment by canavandl** [comment]
What's the "finished rendering" heuristic that you mention?

**Comment by mattpap** [comment]
> What's the "finished rendering" heuristic that you mention?

https://github.com/bokeh/bokeh/blob/master/tests/plugins/phantomjs_screenshot.js#L75. Note this is pretty limited in scope. I'm working on something that will cover more bases (e.g. image loading, maps).

**Comment by bryevdv** [comment]
`plotting/file/bar_chart.py` displays incorrectly to begin (fixes itself after a window resize)

screen shot 2017-04-26 at 08 41 05


A bit more troubling, the `movies` example fails to load, closing the WS connection immediately

**Comment by bryevdv** [comment]
Annoyingly, I if I try to break on all exceptions and run, `movies` works.

**Comment by bryevdv** [comment]
It seems to be working on Chrome and FF but not safari

**Comment by mattpap** [comment]
@bryevdv, everything works fine as much as my testing capabilities allow. I have a suspicion what may not work in safari, but I would need access to it to tests things efficiently.

**Comment by bryevdv** [comment]
OK, well I can look into getting a mac for you to test on. But that will take time.; So my proposal to is spit this PR up. There are lot of changes here that could go into individual PRs:

* changes to settings, embed, client.coffee, etc. 
* style improvements to `get_constraints`
* canvas fixup changes
* renaming constraints to use underscores

Those are probably not the cause of these problems, but they add alot of bulk to this PR. Let's verify that those changes are OK by performing them in isolation, and make the remainder smaller and easier to evaluate.  We can discuss more at the meeting.

**Comment by bryevdv** [comment]
@mattpap  Here is the console screen shot

screen shot 2017-04-26 at 10 40 15

**Comment by mattpap** [comment]
@bryevdv, btw. is this the only failing server example?

**Comment by bryevdv** [comment]
I will check now

**Comment by mattpap** [comment]
`bar_chart` example was broken yesterday. It should be easy to fix. What's worrying is that tests on travis ci didn't pick this up.

**Comment by bryevdv** [comment]
> btw. is this the only failing server example?

Yes, all other examples are working. My only observation is that a couple (surface3d and spectrogram) *seem* to take a bit longer to show up but I'm not 100% certain

**Comment by mattpap** [comment]
> seem to take a bit longer to show up but I'm not 100% certain

That's perfectly possible. I expect minor performance regressions as a cost for correct behavior.

**Comment by mattpap** [comment]
I reviewed all changes to code that may affect the server and there isn't much of that:

* ebf679233bf74bae7fbd5e0e1c8a8bca58230cd1 changes client.coffee, but you said you already tried to revert this, so it's not it.
* f6225b8664328126cc22dbbdc71022d09e9f1135 changes embed.coffee and seems to be a nice candidate for unexpected behavior.

@bryevdv, can you revert the later commit and see if the example works?

**Comment by bryevdv** [comment]
trying now

**Comment by bryevdv** [comment]
@mattpap no that does not solve the problem. However, if I I reduce the data size, even by half:
```
movies = movies[::2]
```
Then it starts working. I believe I may have heard something about safari changing the default max websocket frame size. I thought that was for the future. Perhaps it's not for the future, and somehow is interacting with other changes here. If there is a way to raise that limit we could investigate.

**Comment by mattpap** [comment]
Just to make it clear, this behavior doesn't show up on `master`?

**Comment by bryevdv** [comment]
In fact it is broken on master [1] and is broken (now) on `0.12.5` as well. Does anyone have an old version of safari they can test with? For reference 

safari: 10.1 (12603.1.30.0.34)
OSX: 10.12.4

[1] with the "renders but WS closes and sliders dont work" I described in the call

**Comment by bryevdv** [comment]
OK this def does not have to be fixed here, just the bar chart one. Please make another issue. Also reference this ML discussion:

https://groups.google.com/a/continuum.io/forum/#!searchin/bokeh/websocket$20safari%7Csort:relevance/bokeh/KvrSS3CMBC4/mU0ekn1WAQAJ

I have to step out for an hour or so

ping @canavandl FYI

**Comment by bryevdv** [comment]
6013.1.30 came out a month ago, guessing its the culprit https://en.wikipedia.org/wiki/Safari_version_history

**Comment by mattpap** [comment]
Googling for "safari 10.1 websockets" reveals that we aren't the only unhappy customer.

**Comment by canavandl** [comment]
here's an issue for the safari/websocket problem: https://github.com/bokeh/bokeh/issues/6196

**Comment by bryevdv** [comment]
OK all the relevant layout examples in the repo are working. I'm not sure #4810 or all of #5518 are fixed tho the problems now may be server sychronization. The original example for from #5518 is working, it seems to be adding figures that does not work. I thing there may be an issue adding objects that have references to other instances.

**Comment by bryevdv** [comment]
@mattpap I would support merging this PR as it currently is. There is work to add tests that can probably be rolled into some client work. Are you done / ok to merge?

**Comment by mattpap** [comment]
@bryevdv, lets merge. I still have to finalize making layout reinitializable, so that's why some of the listed issues aren't fully fixed. I will submit a new PR to finish this, today or tomorrow.

**Comment by bryevdv** [comment]
OK, I've updated the top so those issues won't auto close. Merging now.

**Comment by lukauskas** [comment]
With this PR merged, the `document.resize()` function is removed.
I was relying on this function to resize the image from `CustomJS`, an approach similar to the one described [here](http://stackoverflow.com/a/40095304) by birdsarah.

Can someone suggest how to trigger such resize from CustomJS in the new Layout API ?

This is sort-of related to #5381

**Comment by canavandl** [comment]
Thanks for pointing that out @lukauskas , i'm was also relying on ``document.resize`` in another project. 

@mattpap what's the best way to trigger a ``resize`` on a layout with the changes?

**Comment by bryevdv** [comment]
@mattpap this should get a migration note if there is not one already.

**Comment by mattpap** [comment]
@lukauskas, what's described on SO looks like a hack rather than a good practice. Just setting `width` of a figure should do all the necessary resizing, though I doubt this works as expected currently. For now you can use this instead:
```
for (var id in Bokeh.index) {
  Bokeh.index[id].resize()
}
```

**Comment by lukauskas** [comment]
@mattpap oh yes, absolutely, it was definitely a hack.

What I found is that setting `width` (or `height`) is sufficient in practice with the new code.
The caveat is that I do not know any way of telling the layout that the width has changed.
The figure only updates when window is resized.

For my use case, I can afford to automatically trigger the window resize event every time I want to change the size. I am aware though this is as hacky as it gets.

If anyone is interested here's what works:

```python
resize_callback = CustomJS(args=dict(plot=plot),
                                   code="""
                                   var new_height = 12345; // your height here
                                   plot.height = new_height;

                                   // This is the only way to trigger layout re-rendering
                                   // that I can figure out right now
                                   // IE9+ only
                                   var event = document.createEvent('HTMLEvents');
                                   event.initEvent('resize');
                                   window.dispatchEvent(event);
                                   """.format(row_height)
                                   )
```

**Pull request merged**